src/dft/backends/mklcpu/commit.cpp

```diff
@@ -31,113 +31,161 @@
 #include "oneapi/mkl/dft/detail/mklcpu/onemkl_dft_mklcpu.hpp"
 
 #include "oneapi/mkl/dft/detail/commit_impl.hpp"
+
+#include "dft/backends/mklcpu/commit_derived_impl.hpp"
 #include "mkl_service.h"
 #include "mkl_dfti.h"
 
 namespace oneapi {
 namespace mkl {
 namespace dft {
 namespace mklcpu {
+namespace detail {
+
+template <dft::detail::precision prec, dft::detail::domain dom>
+commit_derived_impl<prec, dom>::commit_derived_impl(
+    sycl::queue queue, const dft::detail::dft_values<prec, dom>& config_values)
+        : oneapi::mkl::dft::detail::commit_impl<prec, dom>(queue, backend::mklcpu) {
+    // create the descriptor once for the lifetime of the descriptor class
+    DFT_ERROR status[2] = { DFTI_BAD_DESCRIPTOR, DFTI_BAD_DESCRIPTOR };
 
-template <precision prec, domain dom>
-class commit_derived_impl final : public detail::commit_impl<prec, dom> {
-public:
-    commit_derived_impl(sycl::queue queue, const detail::dft_values<prec, dom>& config_values)
-            : detail::commit_impl<prec, dom>(queue, backend::mklcpu) {
-        DFT_ERROR status = DFT_NOTSET;
+    for (auto dir : { DIR::fwd, DIR::bwd }) {
+        const auto rank = static_cast<std::int64_t>(config_values.dimensions.size());
         if (config_values.dimensions.size() == 1) {
-            status = DftiCreateDescriptor(&handle, get_precision(prec), get_domain(dom), 1,
-                                          config_values.dimensions[0]);
-        }
-        else {
-            status = DftiCreateDescriptor(&handle, get_precision(prec), get_domain(dom),
-                                          config_values.dimensions.size(),
-                                          config_values.dimensions.data());
-        }
-        if (status != DFTI_NO_ERROR) {
-            throw oneapi::mkl::exception(
-                "dft/backends/mklcpu", "commit",
-                "DftiCreateDescriptor failed with status: " + std::to_string(status));
+            status[dir] = DftiCreateDescriptor(&bidirection_handle[dir], mklcpu_prec, mklcpu_dom, 1,
+                                               config_values.dimensions[0]);
+        } else {
+            status[dir] = DftiCreateDescriptor(&bidirection_handle[dir], mklcpu_prec, mklcpu_dom,
+                                               rank,
+                                               config_values.dimensions.data());
         }
     }
 
-    void commit(const detail::dft_values<prec, dom>& config_values) override {
-        set_value(handle, config_values);
-        auto status = DftiCommitDescriptor(handle);
-        if (status != DFTI_NO_ERROR) {
-            throw oneapi::mkl::exception(
-                "dft/backends/mklcpu", "commit",
-                "DftiCommitDescriptor failed with status: " + std::to_string(status));
-        }
+    if (status[0] != DFTI_NO_ERROR || status[1] != DFTI_NO_ERROR) {
```

Comment:
Asking for a confirmation here: is that `||` what we actually want? Don't we want to allow one of them to fail to commit without throwing an exception? Is that exception caught and properly handled if one of the two `status` is not `DFTI_NO_ERROR`?

Reply:
Clarified offline: this related to the creation of the descriptors, so `||` is advisable in this case. Commit-related checks involve `&&`